test/TestKrylov.f90

```diff
@@ -121,6 +123,99 @@ subroutine test_arnoldi_basis_orthogonality(error)
       return
    end subroutine test_arnoldi_basis_orthogonality
 
+   subroutine test_block_arnoldi_full_factorization(error)
+      ! This function checks the correctness of the block Arnoldi implementation by
+      ! verifying that the full factorization is correct, i.e. A @ X[1:p*k] = X[1:p*(k+1)] @ H.
+
+      !> Error type to be returned.
+      type(error_type), allocatable, intent(out) :: error
+      !> Test matrix.
+      class(rmatrix), allocatable :: A
+      !> Krylov subspace.
+      class(rvector), dimension(:), allocatable :: X
+      !> Krylov subspace dimension.
+      integer, parameter :: p = 2
+      integer, parameter :: kdim = test_size/p
+      !> Hessenberg matrix.
+      real(kind=wp) :: H(p*(kdim + 1), p*kdim)
+      !> Information flag.
+      integer :: info
+      !> Misc.
+      integer :: k
+      real(kind=wp) :: Xdata(test_size, p*(kdim + 1))
+      real(kind=wp) :: Rwrk(p,p)
+      double precision, dimension(p*kdim, p*kdim) :: G, Id
+      
+      ! --> Initialize matrix.
+      A = rmatrix(); call random_number(A%data)
+      ! --> Initialize Krylov subspace.
+      allocate (X(1:p*(kdim + 1))); 
+      do k = 1,p
+         call random_number(X(k)%data)
+      enddo
+      call qr_factorization(X(1:p), Rwrk, info)
+      do k = p+1, size(X)
```

Comment:
You can call `initialize_krylov_subspace` right after allocation to zero-out all the vectors before constructing the first p ones. Mostly a cosmetic change getting rid of two lines of code.

Reply:
Correct. Also this is just an adapted copy of the code in `test_arnoldi_full_factorization`. I will change that in both routines